### PR TITLE
PN-7743 - handling of category PREPARE_ANALOG_DOMICILE_FAILURE for timeline events

### PR DESCRIPTION
**Pull request by clombardi**
## Short description
Add the handling of the timeline category PREPARE_ANALOG_DOMICILE_FAILURE to the logic that builds the data structure used to render the timeline in the notification detail page for PF / PG / PA.

## List of changes proposed in this pull request
- Category PREPARE_ANALOG_DOMICILE_FAILURE and the structure of the details for timeline events of this category were added to the model of notifications in pn-commons.
- Class PrepareAnalogDomicileFailureStep added, factory modified to create an instance of this class to generate the text regarding PREPARE_ANALOG_DOMICILE_FAILURE events, in pn-commons.
- Copy for label and description of such events (as specified in https://pagopa.atlassian.net/wiki/spaces/PN/pages/629014723/Timeline+events+e+cosa+mostrare+agli+utenti) added to the `notifiche.json` files for PA / PF / PG.
- The logic of the TimelineStep abstract class was slightly expanded since the physical address related to a timeline event can now be linked to different attributes of the detail, depending of the event category.
- We added tests regarding how the texts for the new event category are computed.

## How to test

> **Note**  
> I defined additional messages to those described in the issue and in https://pagopa.atlassian.net/wiki/spaces/PN/pages/629014723/Timeline+events+e+cosa+mostrare+agli+utenti. These added messages regard scenarios in which the API response does not conform to the specification regarding `failureCause`. I wrote the details in a comment to the issue.

It would be so nice to have examples of notifications including events pertaining to the new category. But the BE-side work won't be completed until (estimatively) the end of the week starting on Monday, September 18th.  
Hence I had to test by a little mocking. To minimize mocking work, I just changed the category and added the specific detail attribute for specific events of existing notifications. This is the code I added to `Notifications.api.ts` for both PA `getSentNotification` and PF `getReceivedNotification`, inside the `if (response.data)` and after the `return` statement there.

```
        if (response.data.iun === 'WZGZ-TDPA-KPHW-202308-H-1') {
          const timelineElementToChange = response.data.timeline.find(elem => elem.elementId === 'SEND_ANALOG_FEEDBACK.IUN_WZGZ-TDPA-KPHW-202308-H-1.RECINDEX_0.ATTEMPT_0');
          if (timelineElementToChange) {
            const detailsToChange = timelineElementToChange.details as any;
            // detailsToChange.foundAddress = detailsToChange.physicalAddress;
            // detailsToChange.failureCause = 'D08';
            detailsToChange.prepareRequestId = detailsToChange.sendRequestId;
            timelineElementToChange.category = TimelineCategory.PREPARE_ANALOG_DOMICILE_FAILURE;
          }
        }
```

I used these notifications in TEST environment

**Mono-recipient for PF and PA**
IUN WZGZ-TDPA-KPHW-202308-H-1 - ada - comune di Palermo
timeline element SEND_ANALOG_FEEDBACK.IUN_WZGZ-TDPA-KPHW-202308-H-1.RECINDEX_0.ATTEMPT_0

**Multi-recipient for PF and PA**
MEKX-DRUP-KXAG-202309-H-1 - ada ed altri - comune di Palermo
timeline element SEND_ANALOG_FEEDBACK.IUN_MEKX-DRUP-KXAG-202309-H-1.RECINDEX_2.ATTEMPT_0

### PG - a deeper mock
For PG the situation is even more delicate, as at the moment is not possible to insert notifications for PG that use the analog delivery; the reason being that a digital address is automatically assigned to any PG and hence digital delivery is used for any  notification whose recipient is a PG.  
In order to perform a test, I decided to transform a digital-flow-related timeline event to a PREPARE_ANALOG_DOMICILE_FAILURE; it looks weird but it works. 

I did the trick for the following notification and element
WEUJ-KYAZ-WHLW-202309-Q-1 - Divina Commedia Srl (utente DanteAlighieri) - comune di Palermo
timeline element SEND_DIGITAL_FEEDBACK.IUN_WEUJ-KYAZ-WHLW-202309-Q-1.RECINDEX_0.SOURCE_GENERAL.REPEAT_false.ATTEMPT_0

This is the code I used in TEST environment; note that I had to define explicitly the physicalAddress.
```
        if (response.data.iun === 'WEUJ-KYAZ-WHLW-202309-Q-1') {
          const timelineElementToChange = response.data.timeline.find(elem => elem.elementId === 'SEND_DIGITAL_FEEDBACK.IUN_WEUJ-KYAZ-WHLW-202309-Q-1.RECINDEX_0.SOURCE_GENERAL.REPEAT_false.ATTEMPT_0');
          if (timelineElementToChange) {
            timelineElementToChange.legalFactsIds = [];
            const detailsToChange = timelineElementToChange.details as any;
            detailsToChange.failureCause = 'D02';
            detailsToChange.foundAddress = {
              "at": "Presso",
              "address": "Via Messi 2022",
              "addressDetails": "SCALA B",
              "zip": "87100",
              "municipality": "MILANO",
              "municipalityDetails": "MILANO",
              "province": "MI",
              "foreignState": "ITALIA"
            };
            detailsToChange.prepareRequestId = 'SEND_DIGITAL.IUN_WEUJ-KYAZ-WHLW-202309-Q-1.RECINDEX_0.SOURCE_GENERAL.REPEAT_false.ATTEMPT_0';
            timelineElementToChange.category = TimelineCategory.PREPARE_ANALOG_DOMICILE_FAILURE;
          }
        }
```

### CODE REVIEW

File: packages/pn-commons/src/utils/TimelineUtils/PrepareAnalogDomicileFailureStep.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        }
+      );
+  
+      if (description.length === 0) {
+        description = getLocalizedOrDefaultLabel(
+          'notifications',
+          `detail.timeline.prepare-analog-domicile-failure-XXX-description${
+            payload.isMultiRecipient ? '-multirecipient' : ''
+          }`,
+          `Non è stato trovato un indirizzo valido per predisporre un altro tentativo di invio - motivo sconosciuto.`,
```
Comment (ZorbaDimatteo) [inline packages/pn-commons/src/utils/TimelineUtils/PrepareAnalogDomicileFailureStep.ts:44]: @clombardi i cannot find this copy here https://pagopa.atlassian.net/wiki/spaces/PN/pages/629014723/Timeline+events+e+cosa+mostrare+agli+utenti was it approved by @AliceAzzolini ?

Comment (AliceAzzolini) [reply packages/pn-commons/src/utils/TimelineUtils/PrepareAnalogDomicileFailureStep.ts:44]: @ZorbaDimatteo I did, you can find in the `PREPARE_ANALOG_DOMICILE_FAILURE` row

File: packages/pn-pa-webapp/public/locales/it/notifiche.json [hunk 2]
```diff
@@ -130,6 +130,17 @@
...
+      "prepare-analog-domicile-failure-D00-description": "Non è stato trovato nessun nuovo indirizzo per predisporre un altro tentativo di invio.",
+      "prepare-analog-domicile-failure-D00-description-multirecipient": "Non è stato trovato nessun nuovo indirizzo per il destinatario {{name}} {{taxId}} per predisporre un altro tentativo di invio.",
+      "prepare-analog-domicile-failure-D01-description": "L'indirizzo {{address}} non ha superato il processo di validazione.",
+      "prepare-analog-domicile-failure-D01-description-multirecipient": "L'indirizzo {{address}} del destinatario {{name}} {{taxId}} non ha superato il processo di validazione.",
+      "prepare-analog-domicile-failure-D02-description": "L'indirizzo {{address}} è già stato usato per effettuare un tentativo di invio.",
+      "prepare-analog-domicile-failure-D02-description-multirecipient": "L'indirizzo {{address}} del destinatario {{name}} {{taxId}} è già stato usato per effettuare un tentativo di invio.",
+      "prepare-analog-domicile-failure-XXX-description": "Non è stato trovato un indirizzo valido per predisporre un altro tentativo di invio - motivo sconosciuto {{failureCause}}.",
```
Comment (ZorbaDimatteo) [inline packages/pn-pa-webapp/public/locales/it/notifiche.json:140]: @clombardi same for these copies

File: packages/pn-commons/src/utils/TimelineUtils/PrepareAnalogDomicileFailureStep.ts
Comment (ZorbaDimatteo) [inline packages/pn-commons/src/utils/TimelineUtils/PrepareAnalogDomicileFailureStep.ts:1]: @clombardi is it possible to add some tests to this new step?

**Review by ZorbaDimatteo — CHANGES_REQUESTED** [review]
@clombardi please review some copies with @AliceAzzolini and if possible generate some tests for the new added step (since they have been added during test refactoring for pn-commons)

File: packages/pn-commons/src/utils/TimelineUtils/PrepareAnalogDomicileFailureStep.ts (hunk 1 above)
Comment (ZorbaDimatteo) [reply packages/pn-commons/src/utils/TimelineUtils/PrepareAnalogDomicileFailureStep.ts:44]: @AliceAzzolini I can't find this specific copy "Non è stato trovato un indirizzo valido per predisporre un altro tentativo di invio - motivo sconosciuto." I suppose is used to manage a default error

**Review by ZorbaDimatteo — APPROVED** [review]

**Review by ZorbaDimatteo — APPROVED** [review]

**Pull request merged**